osquery/tables/system/windows/registry.cpp

```diff
@@ -278,33 +279,150 @@ void queryKey(const std::string& keyPath, QueryData& results) {
     delete[](bpDataBuff);
   }
   RegCloseKey(hRegistryHandle);
-};
+}
 
-QueryData genRegistry(QueryContext& context) {
-  QueryData results;
-  std::set<std::string> rKeys;
-  auto shouldWarnLocalUsers = false;
-  /// By default, we display all HIVEs
-  if ((context.constraints["key"].exists(EQUALS) &&
-       context.constraints["key"].getAll(EQUALS).size() > 0)) {
-    rKeys = context.constraints["key"].getAll(EQUALS);
-    shouldWarnLocalUsers = true;
+static inline void populateDefaultKeys(std::set<std::string>& rKeys) {
+  boost::copy(kRegistryHives | boost::adaptors::map_keys,
+              std::inserter(rKeys, rKeys.end()));
+}
+
+static inline void replaceKeysWithSubkeys(std::set<std::string>& rKeys) {
+  std::set<std::string> newKeys;
+  for (const auto& key : rKeys) {
+    QueryData regResults;
+    queryKey(key, regResults);
+    for (const auto& r : regResults) {
+      if (r.at("type") == "subkey") {
+        newKeys.insert(r.at("path"));
+      }
+    }
+  }
+  rKeys = newKeys;
+}
+
+static inline void appendSubkeyToKeys(const std::string& subkey,
+                                      std::set<std::string>& rKeys) {
+  std::set<std::string> newKeys{};
+  for (auto& key : rKeys) {
+    newKeys.insert(key + kRegSep + subkey);
+  }
+  rKeys = std::move(newKeys);
+}
+
+static inline Status populateAllKeysRecursive(
+    std::set<std::string>& rKeys,
+    size_t currDepth = 1,
+    size_t maxDepth = kRegMaxRecursiveDepth) {
+  std::set<std::string> subkeys{};
+
+  if (currDepth > maxDepth) {
+    return Status(1, "Max recursive depth reached");
+  }
+
+  for (const auto& key : rKeys) {
```

Comment:
Not quite, replaceKeysWithSubkeys replaces the original set with the subkeys, whereas this appends the subkeys and preserves the parent. However, this raises a good point, the code is similar enough that I shouldn't recreate it.